Standard user account

jannikluhn opened the pull request:
The standard user account is a contract parametrized by an address that
- Validates that the transaction is signed with the private key corresponding to the address,
- validates that the block number is in a certain range (given by the transaction),
- stores a nonce that is incremented at every call,
- pays for gas, and
- calls another contract.

It also has additional methods to
- extract the nonce and to
- validate an arbitrary signed message against the owner address.

Once the PR is finished it should contain
- the contract code both as Vyper LLL and compiled bytecode
- extensive tests for the contract code
- utility functions to construct transactions in a format understood by the contract
- utility functions to deploy such a contract (which might be split into a main and a validation contract)

References:
- https://ethresear.ch/t/tradeoffs-in-account-abstraction-proposals/263
- https://ethresear.ch/t/validation-code-for-abstraction/335/9

pipermerriam commented on evm/auxiliary/user_account.py:
```diff
@@ -0,0 +1,181 @@
+    #
+
+    @property
+    def signature(self):
```
what do you think about renaming this to `vrs` instead of `signature`?  I think that naming convention would match better with other portions of both this codebase and others (web3.py and eth-keys).

With the conventions we've been establishing, I would expect this to return a `eth_keys.Signature` object.

jannikluhn replied:
Do you prefer renaming it to `vrs` or making it return a `Signature` object? I have no preference in that regard.

pipermerriam replied:
I think `vrs` is my preference since this is a low-level API

pipermerriam commented on evm/auxiliary/user_account.py:
```diff
@@ -0,0 +1,181 @@
+        self.data = self.data[:VALUE_SLICE.start] + b + self.data[VALUE_SLICE.stop:]
+
+    @property
+    def receiver(self):
```
This fragments the terminology that we use for the *destination* of a transaction.  Elsewhere we're using `to`.  Any thoughts on whether it'd be better to conform to this naming convention?

jannikluhn replied:
Not sure... We used to have `to` and `sender`. Now we have `target`, `sender` and `receiver`.

- `to` and `target` refer to more or less the same thing I guess (the first account that is called)
- `sender` changes meaning: it used to be the one who signed the transaction, now it's the "owner" of the `target` account (and just happens to put a signature in the `data` field)
- `receiver` is the second hop (the account that's called by the `target` account)

So from a technical standpoint, I don't see a lot of overlap between original and new terms. I could imagine renaming `target` to `to` for consistency (although I'd prefer `target` in a vacuum). Renaming `receiver` to `to` could be misleading on the protocol level, but would make sense from a high level point of view in most cases. I don't see a good alternative to `sender`.

pipermerriam replied:
Good explanation.  I think you are right to leave this as-is and not change the names from `target/sender/receiver`

pipermerriam replied:
Does the name `destination` better capture the concept that `receiver` currently represents?

pipermerriam commented on evm/logic/context.py:
```diff
@@ -176,3 +176,7 @@ def returndatacopy(computation):
+
+
+def sighash(computation):
+    pass
```
thoughts on changing this to a `raise NotImplementedError`?

pipermerriam commented on evm/logic/system.py:
```diff
@@ -173,3 +173,7 @@ def __call__(self, computation):
+
+
+def paygas(computation):
+    pass
```
thoughts on changing this to a `raise NotImplementedError`?

pipermerriam commented on evm/vm/forks/sharding/opcodes.py:
```diff
@@ -0,0 +1,33 @@
+        mnemonic=mnemonics.PAYGAS,
+        gas_cost=0,
+    ),
+
```
whitespace

pipermerriam commented on tests/auxiliary/user-account/test_user_account.py:
```diff
@@ -0,0 +1,172 @@
+
+def test_valid_transfer(chain, tx1):
+    vm = chain.get_vm()
+    vm.apply_transaction(tx1)
```
Are there any assertions that we could make here?

jannikluhn replied:
Definitely. That's one of the reasons why it's still WiP (this test doesn't even pass yet)

pipermerriam commented on evm/auxiliary/user_account.py:
```diff
@@ -0,0 +1,181 @@
+
+
+USER_ACCOUNT_CODE = decode_hex(
+    "0x3f595260805903608020600052602060006080600060006001610bb8f1506001600051"
```
Will you add a comment referencing the *source code* that produced this bytecode and how it can be reproduced/verified.

jannikluhn replied:
Yes, this is not much more than a placeholder for now. Ideally I'd like to put the source code there directly and compile it in a test for comparison, but this will probably not work because Viper needs Python 3.6.

hwwhww replied:
#206 has the same issue, it's better not import Viper even in Python 3.6. 

A (little complicated) suggestion of verifying Viper code:
1. A command for compiling Viper source code to bytecode.
2. If someone want to update the code, they have to upload `source code` and `bytecode` (can be in separate file or assigned to a global variable in this case)
3. Install Viper in CI testing.
4. Write a test for auto-compiling the source code and then comparing the compiled bytecode with `evm.auxiliary.user_account.USER_ACCOUNT_CODE`.

By the way, @pipermerriam what do you think about upgrading to Python 3.6?

jannikluhn replied:
Perhaps the cleanest solution would be to put the code plus tests in a separate repository. The repository would contain packages with the bytecode for different languages that can be imported into client implementations such as Py-EVM.

pipermerriam replied:
> By the way, @pipermerriam what do you think about upgrading to Python 3.6?

I'm in favor: https://github.com/ethereum/py-trie/pull/18

pipermerriam replied:
I'm indifferent to the actual solution for the bytecode creation as long as it's documented and independently verifiable.

jannikluhn commented:
I've given up on testing the contract code for now, as it's not really possible at the moment. I'll come back to it once #238 is done, but for now I'd like to put this PR on hold.

pipermerriam commented:
I think that a large part of this diff will go away once this is rebased.

jannikluhn commented:
Updated the description. Sorry for the current state of the PR. Now that #238 is merged I'm back at writing tests/debugging the contract. In the meantime I've made a couple of changes in the contract. I'm going to push something later today.

jannikluhn commented:
Still not completely finished, but almost there I hope.

To run the tests, cherry-pick https://github.com/ethereum/py-evm/pull/259/commits/90d155d23ef95d18c24170e168414b634e9b7481, resolve the merge conflicts (accept both), and replace [this line](https://github.com/jannikluhn/py-evm/blob/6e76c1277a82fa6b2241501ecc5d71630fec205d/evm/logic/system.py#L196) with `pass`.

pipermerriam commented:
Sighash is now merged so this can be rebased (as I assume it needs sighash in it to work)

pipermerriam commented:
eh, should have said sighash is *ready for merge* :P

jannikluhn commented:
Merged the sighash PR and rebased. This should be ready for review now.

I'm not particularly happy with some names in this one, especially `ForwardingTransaction`. I called it `forwarding` because the "payload" data is passed through to the actual "destination" contract. But it's specific to one particular user contract, others might require a different format.

Test-wise I'd like to see tests that make sure that everything works as expected even if the gas runs out at different points during the execution of the contract. But this can only be done once PAYGAS is properly implemented.

jannikluhn commented:
Ooh, I missed one thing: `access_list` is left empty all the time. It should default to the whole storage space of the destination account and to the required parts of user and validation accounts (as suggested [here](https://github.com/ethereum/py-evm/issues/194#issuecomment-356503391). Tests are not failing because of #263 .

pipermerriam commented on evm/auxiliary/user_account_contract/transaction.py:
```diff
@@ -0,0 +1,210 @@
+EMPTY_DATA = b'\x00' * DESTINATION_SLICE.stop
+
+
+class ForwardingTransaction(ShardingTransaction):
```
Other name ideas

- `StandardAccountTransaction`
- `ProxyTransaction`
- `EIP86Transaction` (or whatever the appropriate EIP number is)

jannikluhn replied:
`StandardAccountTransaction` was the name before I switched to `ForwardingTransaction`. I dismissed it because I'm not sure if "standard account" is a good term either. It should probably be rather called "standard user account", and even that assumes that this contract will actually become "the" standard. But `StandardUserAccountTransaction` strikes me as too long.

General note on EIPs: As far as I know there are no EIPs for the sharding protocol itself. At least at the moment, I don't know if that will change.

pipermerriam commented on evm/auxiliary/user_account_contract/transaction.py:
```diff
@@ -0,0 +1,210 @@
+        self.max_block = max_block
+        self.nonce = nonce
+        self.int_gas_price = gas_price
+        self.msg_data = msg_data
```
It looks like this class needs to implement a `validate()` method to validate these additional parameters.

pipermerriam commented on evm/auxiliary/user_account_contract/transaction.py:
```diff
@@ -0,0 +1,210 @@
+            self.sig_hash,
+        ])
+
+    def sign(self, private_key):
```
Was it intentional to deviate from the existing API for transaction signing which uses `as_signed_transaction` which returns a new object rather than mutating the existing object?

This question is from a broader question about the use of property setters.

- **if** the use of property setters was to embed validation as part of setting the value, then I would suggest changing to use a single `validate()` method which is inline with the other transaction classes.
- **if** the use of property setters was to allow mutation of the object, I would like to understand why since we've trended away from most forms of object mutation.

jannikluhn replied:
I added the setters to allow mutation. The other transaction objects are technically mutable as well and I wanted to mirror that. But if they aren't mutated in practice I'll be happy to remove them because all these setters look quite ugly. I'll also try to implement the `as_signed_transaction` pattern.

pipermerriam commented on tests/auxiliary/user-account/test_contract.py:
```diff
@@ -0,0 +1,372 @@
+    "code": b"",
+}
+
+DEFAULT_TX_PARAMS = {**DEFAULT_BASE_TX_PARAMS, **{
```
This method of dictionary merging has always bugged me, mostly because it takes advantage of the implementation rather than using formal semantics or which keys take precedence.  I admit it could be argued that the way key precedence is handled in this case is part of the formal semantics for how this works, but either way, it's always felt a bit like magic.

What do you think about making use of something like the following.  (If you haven't figured it out by now, I have great love for the `toolz/cytoolz` library)

```python
from cytoolz import (
    merge,
    dissoc,
)

DEFAUT_TX_PARAMS = merge(
    dissoc(DEFAULT_BASE_TX_PARAMS, 'code'),
    {...<the overrides>},
}
```

Zero mutation and explicit about dictionary merging.

jannikluhn replied:
Personally I don't prefer one way over the other, so I don't mind adopting this style.

pipermerriam commented on tests/auxiliary/user-account/test_contract.py:
```diff
@@ -0,0 +1,372 @@
+        (v, r, s - 1),
+        (v, r, 0),
+
+        (27 if v == 28 else 28, r, SECPK1_N - s)
```
can we parametrize these using `pytest.mark.parametrize` so that they each run in isolation?

pipermerriam commented on tests/auxiliary/user-account/test_contract.py:
```diff
@@ -0,0 +1,372 @@
+
+    logs = computation.get_log_entries()
+    assert len(logs) == 1
+    assert big_endian_to_int(logs[0][-1]) > 900 * 1000  # some gas will have been consumed earlier
```
A slightly easier to understand version of this assertion that I've used elsewhere is to assert that the delta is smaller than a certain value.

```
gas_used = big_endian_to_int(logs[0][-1])
assert transaction.gas - gas_used < 1000
```

hwwhww commented on tests/auxiliary/user-account/test_contract.py:
```diff
@@ -0,0 +1,372 @@
+        block_number=10,
+        gas_limit=3141592,
+        timestamp=1422494849,
+        parent_hash=decode_hex("0000000000000000000000000000000000000000000000000000000000000000"),
```
You can use `evm.constants.ZERO_HASH32`.

pipermerriam commented on evm/auxiliary/user_account_contract/transaction.py:
```diff
@@ -71,18 +61,98 @@ def __init__(
             code=b'',
         )
+
+    def create_unsigned_transaction(self):
```
To stick with the existing transaction API this should be a `classmethod`.  This API exists to reduce the cumbersome import of both the signed and unsigned classes, allowing you to create unsigned transactions from the main transaction class.  I'd be fine with this just being an `*args, **kwargs` passthrough to reduce size since there are 10+ arguments.

pipermerriam commented on tests/auxiliary/user-account/test_contract.py:
```diff
@@ -282,91 +286,107 @@ def test_call_checks_signature(vm):
+            s=s,
+        )
+        # don't detect invalidity before transaction execution
+        transaction.validate = lambda: None
```
I think a more appropriate way to check this would be to assemble a `Message` and then to run `vm.get_computation(message).apply_message()`.

This should execute the transaction *below* the place where the call to `transaction.validate()` occurs, allowing you to directly test that the EVM code appropriately validates the signature without having to do any monkeypatching of the transaction object.

hwwhww commented:
I was checking with @vbuterin  about renaming some parameters and also mentioned about 
`ShardingTransaction.to` or `ShardingTransaction.target`, he clarified his intention:
> so the reason I wanted to rename it to target is that in real life, the transaction.to field will be the sender of the money (and/or function call)
as it would be the forwarding contract that controls the account from which the funds get sent and the receiver would be specified in the tx.data under some encoding
I think it would cause too much long-term risk of devs getting confused for `transaction.to` to be the sender
hence my idea to rename it `target`

jannikluhn commented:
Addressed comments and rebased.

@vbuterin Do you have time to review the contract code? https://github.com/ethereum/py-evm/pull/230/files#diff-05178e170e27782a10a1498ec2aefdef

vbuterin commented on evm/auxiliary/user_account_contract/contract.py:
```diff
@@ -0,0 +1,211 @@
+        ['if',
+            ['eq',
+                ['calldataload', CALLDATA_FUNCTION_ID],
+                NONCE_GETTER_ID],
```
Shouldn't NONCE_GETTER_ID be multiplied by 2^224? It's the first four bytes that are the function sig, so getting the first 32 bytes would return that with 28 zero bytes at the end.

BTW are there tests for all of this? If so, the tests should catch this.

jannikluhn replied:
You're right, will fix this.

Tests are here: https://github.com/jannikluhn/py-evm/blob/07587ac50734c62a7d8862c072fbcaeedaa93c97/tests/auxiliary/user-account/test_contract.py

They cheat a bit by importing the function ids, that's why they didn't catch this. Will add a test comparing against a manually generated one (using the Vyper code).

vbuterin commented on evm/auxiliary/user_account_contract/contract.py:
```diff
@@ -0,0 +1,211 @@
+                ['calldataload', CALLDATA_FUNCTION_ID],
+                VALIDATION_CODE_GETTER_ID,
+                # return validation code address
+                ['mstore', 0, validation_code_address],
```
Now that I think about it, what *would* the validation code address be? We don't want to create two contracts per account. I would actually recommend just not bothering with this `get_validation_code` feature at this point to keep it simpler.

vbuterin commented on evm/auxiliary/user_account_contract/contract.py:
```diff
@@ -0,0 +1,211 @@
+                # output data (discarded)
+                0,
+                0
+            ]],
```
Let's not do things this way. This would require calling 2 accounts to send from this account, which would blow up witness size by a factor of 2. IMO just do ecrecover here.

vbuterin commented on evm/auxiliary/user_account_contract/contract.py:
```diff
@@ -0,0 +1,211 @@
+        ['sstore', STORAGE_NONCE, ['add', ['calldataload', CALLDATA_NONCE], 1]],
+
+        # Assert that we won't run out of gas from here on
+        ['assert', ['gt', 'gas', GAS_RESERVE + GAS_RESERVE_OFFSET]],
```
Is this line still necessary?

jannikluhn replied:
Depends: If we run out of gas after the PAYGAS call, the transaction can be replayed infinitely and the account would get drained. This line makes sure that this can never happen. If we trust the user to never create transactions with not enough gas, the check isn't necessary. But at least I would feel much safer with it.

jannikluhn commented:
Anything else or can we merge?

pipermerriam commented:
I can get onboard with the rename from `to` to `target` with that explanation.

jannikluhn commented:
I'm also in favor of renaming `to` to `target`, but I think this should be done in a separate pull request.

jannikluhn commented:
Rebased. Is this `Ready-for-merge` when tests have passed?

hwwhww commented on evm/auxiliary/user_account_contract/contract.py:
```diff
@@ -0,0 +1,157 @@
+VALIDATION_CODE_GAS = 3500
+GAS_RESERVE = 4500  # amount of gas reserved for returning
+GAS_RESERVE_OFFSET = 200
+NONCE_GETTER_ID = 0x141b5b48
```
maybe add a comment to describe
`# NONCE_GETTER_ID = big_endian_to_int(keccak(b"get_nonce()")[:4])` ?

hwwhww commented:
They are green now  🎉
